### PR TITLE
Toimage bug

### LINKED ISSUE #600: Plotly.toImage and Plotly.downloadImage not working correctly
It appears to work on the most recent Ubuntu version of Chrome, but on  OSX `Google Chrome v51.0.2704.79`, things are no bueno.

cc @timelyportfolio

### PR DESCRIPTION
**Pull request by mdtusz**
Fixes #600.

This PR fixes a nasty little bug with image downloads. The culprit was font-family strings that were quoted and breaking the svg string used for image conversions.

The fix is not exactly _pretty_, but should be more robust now (previously, the quotation marks were escaped _before the svg was converted to a string_ - Chrome seems to not care for that and automatically replaces the double quotation marks.

### CODE REVIEW

File: src/snapshot/svgtoimg.js [hunk 1]
```diff
@@ -86,7 +86,7 @@ function svgToImg(opts) {
                     imgData = canvas.toDataURL('image/webp');
                     break;
                 case 'svg':
-                    imgData = svg;
+                    imgData = url;
```
Comment (mdtusz) [inline src/snapshot/svgtoimg.js]: The `imgData` must be an actual URL to work as a download via a link.

File: src/snapshot/tosvg.js [hunk 2]
```diff
@@ -78,24 +78,12 @@ module.exports = function toSVG(gd, format) {
...
-             *
-             * for more info.
-             */
+            // Font family styles break things because of quotation marks,
+            // so we must remove them *after* the SVG DOM has been serialized
+            // to a string (browsers convert singles back)
             var ff = txt.style('font-family');
             if(ff && ff.indexOf('"') !== -1) {
-                txt.style('font-family', ff.replace(/"/g, '\\\''));
+                txt.style('font-family', ff.replace(/"/g, 'TOBESTRIPPED'));
```
Comment (mdtusz) [inline src/snapshot/tosvg.js]: We can switch this to something even less likely to have collisions if we think it's necessary.

File: src/snapshot/svgtoimg.js (hunk 1 above)
Comment (etpinard) [inline src/snapshot/svgtoimg.js]: good to know!

File: test/jasmine/tests/download_test.js [hunk 3]
```diff
@@ -44,45 +44,84 @@ describe('Plotly.downloadImage', function() {
...
+
+// Only chrome supports webp at the time of writing
+function checkWebp(cb) {
+    var img = new Image();
+    img.onload = function() {
+        cb(true);
+    };
+
+    img.onerror = function() {
+        cb(false);
```
Comment (etpinard) [inline test/jasmine/tests/download_test.js]: smooth.

File: src/snapshot/tosvg.js (hunk 2 above)
Comment (etpinard) [inline src/snapshot/tosvg.js]: I'm ok with this. The most important part now is that all our supported image types are tested.

Down the road, we might want to put these plotly-specific constants in a separate well-documented file (along with e.g. [`Axed.BADNUM`](https://github.com/plotly/plotly.js/blob/19a37357341e7e098ce2c20ee98f65965c2c5d66/src/plots/cartesian/constants.js#L30)). But, there's no need to do this now.

 :+1:

**Comment by etpinard** [comment]
Great. @mdtusz Can you try this patch on browserstack (if you haven't already) ?

I think the easiest way is:

- `npm run build`
- commit the dist folder to a branch and push
- write up a codepen using src:

```
https://rawgit.com/plotly/plotly.js/blob/<name-of-branch>/dist/plotly.min.js
```

- select a few browsers and run a screenshot test.

**Comment by mdtusz** [comment]
Looks all good for Chrome, FF, Safari, and IE11! 🎉

**Comment by etpinard** [comment]
:dancer:

**Comment by kimli42** [comment]
From which version of Safari does this work? I have plotly v.1.23.0 and downloading images (scattergl plot specifically is what I tested) doesn't work in Safari 8.

**Comment by rreusser** [comment]
For reference: Safari 8 was released on Oct 16, 2014. Current is Safari 10. Are you able to download this plot? https://plot.ly/~chris/17389.embed It works for me in Safari 10.

**Comment by kimli42** [comment]
Unfortunately, you can't upgrade to the latest Safari (on Mac) without upgrading the OS. That is why I was asking if anyone else was able to use this feature in Safari 8.

**Comment by rreusser** [comment]
Is there a specific error in the console? There is already [some code to handle specific browsers](https://github.com/plotly/plotly.js/pull/446/files#diff-4dfce3f7c28b0dba495e2d9b02fcdf52R35), so it's definitely a tricky issue to get working across all browsers past and present.

**Comment by kimli42** [comment]
There's no error in the console. I've encountered this before trying to developing a file downloader in legacy Safari.

**Comment by gschivley** [comment]
I've seen the closed issue #250, but can't get .png downloads to work with the latest plotly and Safari 10.0.

**Pull request merged**